lib/index.js

```diff
@@ -1,3 +1,232 @@
+            const oldRequired = oldSubSchema.required && oldSubSchema.required.includes(paramName);
+            const newRequired = newSubSchema.required && newSubSchema.required.includes(paramName);
+            const keyAdded = utils.keyAdded(
+                oldSubSchema.properties,
+                newSubSchema.properties,
+                paramName
+            );
+            const keyRemoved = utils.keyRemoved(
+                oldSubSchema.properties,
+                newSubSchema.properties,
+                paramName
+            );
+
+            switch (context.compType) {
+                case FORWARD:
+                    // DISALLOWED to make a required property optional.
+                    if (!newRequired && oldRequired) {
+                        return [ error('Not required anymore') ];
+                    }
+
+                    if (keyAdded) {
+                        // ALLOWED to add both required and optional properties
+                        return [];
+                    } else if (keyRemoved) {
+                        if (oldRequired) {
+                            // DISALLOWED to remove required properties
+                            return [ error('Required field removed') ];
+                        }
+                        // ALLOWED to remove optional properties
+                        return [];
+                    }
+                    break;
+                case BACKWARD:
+                    // DISALLOWED to make an property required
+                    if (newRequired && !oldRequired) {
+                        return [ error('Required now') ];
+                    }
+
+                    if (keyRemoved) {
+                        // ALLOWED to delete both required and optional properties
+                        return [];
+                    } else if (keyAdded) {
+                        // DISALLOWED to add required properties
+                        if (newRequired) {
+                            return [ error('Required field added') ];
+                        }
+                        // ALLOWED to add optional properties
+                        return [];
+                    }
+                    break;
+            }
+
+            return applyChecks(
+                oldSubSchema.properties[paramName],
+                newSubSchema.properties[paramName],
+                newContext
+            );
+        }).reduce((arr1, arr2) => arr1.concat(arr2));
+    },
+    string: (oldSubSchema, newSubSchema, context) => {
+        utils.verifySchemasType('string', oldSubSchema, newSubSchema);
+        // TODO: figure out string-related forward/backward compat rules.
+        return [];
+    }
+};
+
+/**
+ * TODO
+ * @param oldSubSchema
+ * @param newSubSchema
+ * @param context
+ */
+const applyChecks = (oldSubSchema, newSubSchema, context) => {
+    const path = context.path;
+    if (!oldSubSchema.type && !newSubSchema.type) {
```

Comment:
This really depends on whether a prop is being added or removed (or its type), doesn't it?